### ISSUE TITLE
Support passing functions in the default logger implementations

### ISSUE THREAD

**Issue by venkatd** [leisim/logger#45]
Some of our logger statements involve expensive computations like formatting objects to make the logs easy to read. To work around this, we are passing functions for the expensive log statements and they are getting evaluated in the printer right before they need to be printed.

For example:
```dart
    logger.d(() =>
        'SEND ref=$ref topic=$topic event=$event payload=' +
        summarizeGqlPayload(payload).toString());
```

`summarizeGqlPayload` in this case is relatively expensive and something we wouldn't want to be execute in production. Or at least all these little things add up to increased CPU usage. Hence we are passing a function.

Our corresponding `stringify` in our custom logger looks something like this to handle functions.

```dart
  String stringifyMessage(dynamic message) {
    final finalMessage = message is Function ? message() : message;
    if (finalMessage is Map) {
      return _encoder.convert(finalMessage);
    } else if (finalMessage is Iterable) {
      return _encoder.convert(finalMessage);
    } else {
      return finalMessage.toString();
    }
  }
```

Is this worth having as a default for the logger?

**Comment by mercutiodesign** [comment]
I was just looking for exactly this functionality and I think it would be a nice enhancement.

**Comment by venkatd** [comment]
@mercutiodesign for now your best bet would be to implement your own printer by copying/pasting the closest one to what you need. Can check if the message is a function and evaluate it before you print.

**Comment by haarts** [comment]
I rather like the idea! Since the `message` argument to a log call (like `d`) is `dynamic` already this should be relatively straightforward.

**Comment by smotastic** [comment]
Hi, needed exactly the same but saw that there is no PR for this, so I just did it.
See #105

**Comment by haarts** [comment]
Closed since #105 is merged.
